### ISSUE TITLE
[torch.compile] the sum of `softmax` isn't `1` on cuda

### ISSUE THREAD

**Issue by payphon** [pytorch/pytorch#101039]
### 🐛 Describe the bug

`torch.compile` returns the sum of `softmax` isn't equal to `1` on cuda

```py
import torch

torch.manual_seed(420)

class Model(torch.nn.Module):

    def __init__(self):
        super().__init__()
        self.softmax = torch.nn.Softmax(dim=-1)

    def forward(self, query, key):
        y = torch.matmul(query, key.transpose(-2, -1))
        z = y.div(1e-06)
        w = self.softmax(z)
        return w

device = 'cuda'
func = Model().to(device)

query = torch.randn(1, 10, 40).to(device)

key = torch.randn(1, 2, 40).to(device)

jit_func = torch.compile(func)

res1 = func(query, key) # without jit
print(res1)
# tensor([[[0., 1.],
#          [0., 1.],
#          [0., 1.],
#          [0., 1.],
#          [0., 1.],
#          [1., 0.],
#          [1., 0.],
#          [0., 1.],
#          [0., 1.],
#          [0., 1.]]], device='cuda:0')

res2 = jit_func(query, key)
print(res2)
# tensor([[[0.0000, 0.9804],
#          [0.0000, 1.0561],
#          [0.0000, 1.1190],
#          [0.0000, 0.9721],
#          [0.0000, 1.1999],
#          [0.9972, 0.0000],
#          [0.8769, 0.0000],
#          [0.0000, 1.0315],
#          [0.0000, 1.0398],
#          [0.0000, 0.8039]]], device='cuda:0')
```

### Versions

<details>
<summary>Click to expand</summary>

```
Collecting environment information...
PyTorch version: 2.1.0.dev20230503+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.1 LTS (x86_64)
GCC version: (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0
Clang version: 14.0.0-1ubuntu1
CMake version: Could not collect
Libc version: glibc-2.35

Python version: 3.9.16 (main, Mar  8 2023, 14:00:05)  [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.19.5-051905-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 11.5.119
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3060
Nvidia driver version: 510.108.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   46 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          24
On-line CPU(s) list:             0-23
Vendor ID:                       GenuineIntel
Model name:                      12th Gen Intel(R) Core(TM) i9-12900K
CPU family:                      6
Model:                           151
Thread(s) per core:              2
Core(s) per socket:              16
Socket(s):                       1
Stepping:                        2
CPU max MHz:                     6700.0000
CPU min MHz:                     800.0000
BogoMIPS:                        6374.40
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault cat_l2 invpcid_single cdp_l2 ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdt_a rdseed adx smap clflushopt clwb intel_pt sha_ni xsaveopt xsavec xgetbv1 xsaves split_lock_detect avx_vnni dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req hfi umip pku ospke waitpkg gfni vaes vpclmulqdq tme rdpid movdiri movdir64b fsrm md_clear serialize pconfig arch_lbr ibt flush_l1d arch_capabilities
Virtualization:                  VT-x
L1d cache:                       640 KiB (16 instances)
L1i cache:                       768 KiB (16 instances)
L2 cache:                        14 MiB (10 instances)
L3 cache:                        30 MiB (1 instance)
NUMA node(s):                    1
NUMA node0 CPU(s):               0-23
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Retbleed:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected

Versions of relevant libraries:
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.24.2
[pip3] pytorch-triton==2.1.0+7d1a95b046
[pip3] torch==2.1.0.dev20230503+cu118
[pip3] torchaudio==2.1.0.dev20230503+cu118
[pip3] torchvision==0.16.0.dev20230503+cu118
[conda] numpy                     1.24.2                   pypi_0    pypi
[conda] pytorch-triton            2.1.0+7d1a95b046          pypi_0    pypi
[conda] torch                     2.1.0.dev20230503+cu118          pypi_0    pypi
[conda] torchaudio                2.1.0.dev20230503+cu118          pypi_0    pypi
[conda] torchvision               0.16.0.dev20230503+cu118          pypi_0    pypi
```

</details>

cc @ezyang @msaroufim @wconstab @ngimel @bdhirsh @anijain2305 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @soumith

**Comment by ezyang** [comment]
@ngimel was looking at this and telling me it's due to fma

**Comment by ngimel** [comment]
Also, the blast radius of this bug is very small - it happens only with unrolled reductions (reduction_dim <=8) and values are meaningfully different only if they are very large (that's why `.div(1e-6)` is needed. We still might want to disable unrolled reduction for amax/amin to avoid this particular pattern, but in general, we cannot guarantee that underneath compiler won't use fma instead of a*b+c, and that could bring some small difference with it.

**Comment by peterbell10** [comment]
> we cannot guarantee that underneath compiler won't use fma instead of a*b+c, 

 If we generate `tl.math.div_rn(a, b)` then the rounding mode is respected. Though in general it may of course perform worse. Here is the softmax kernel for example:

```python
In [2]: import torch
   ...: from torch._inductor import triton_helpers
   ...: import triton
   ...: import triton.language as tl
   ...: 
   ...: @triton.jit
   ...: def triton_(in_ptr0, out_ptr0, xnumel, XBLOCK : tl.constexpr):
   ...:     xnumel = 8
   ...:     xoffset = tl.program_id(0) * XBLOCK
   ...:     xindex = xoffset + tl.arange(0, XBLOCK)[:]
   ...:     xmask = xindex < xnumel
   ...:     x2 = xindex
   ...:     x1 = (xindex // 2)
   ...:     tmp0 = tl.load(in_ptr0 + (x2), xmask)
   ...:     tmp3 = tl.load(in_ptr0 + (2*x1), xmask)
   ...:     tmp5 = tl.load(in_ptr0 + (1 + (2*x1)), xmask)
   ...:     tmp1 = 1e-06
   ...:     tmp2 = tl.math.div_rn(tmp0, tmp1)
   ...:     tmp4 = tl.math.div_rn(tmp3, tmp1)
   ...:     tmp6 = tl.math.div_rn(tmp5, tmp1)
   ...:     tmp7 = triton_helpers.maximum(tmp4, tmp6)
   ...:     tmp8 = tmp2 - tmp7
   ...:     tmp9 = tl.exp(tmp8)
   ...:     tmp10 = tmp4 - tmp7
   ...:     tmp11 = tl.exp(tmp10)
   ...:     tmp12 = tmp6 - tmp7
   ...:     tmp13 = tl.exp(tmp12)
   ...:     tmp14 = tmp11 + tmp13
   ...:     tmp15 = tmp9 / tmp14
   ...:     tl.store(out_ptr0 + (tl.broadcast_to(x2, [XBLOCK])), tmp15, xmask)
   ...: 
   ...: arg = torch.rand(4, 2, device="cuda")
   ...: buf = torch.empty(4, 2, device="cuda")
   ...: triton_[(1,)](arg, buf, 8, 8)
   ...: print(buf)
tensor([[0., 1.],
        [1., 0.],
        [0., 1.],
        [0., 1.]], device='cuda:0')

```

**Comment by ngimel** [comment]
Hm, maybe we can replace divisions by a scalar by multiplication - it seems to help softmax case, although `x-x.max()` still doesn't produce exact 0s with multiplication.

**Comment by Chillee** [comment]
Going to assign to @peterbell10 for now.

**Comment by tugsbayasgalan** [comment]
@peterbell10 are you still planning to work on it?
